File: 0678-valid-parenthesis-string/0678-valid-parenthesis-string.py
class Solution:
    def checkValidString(self, s: str) -> bool:
        open_count= 0
        for c in s:
            if c == '(' or c == '*':
                open_count += 1
            else:
                open_count -= 1

            if open_count < 0:  # If at any point, more ')' than '(' or '*' are encountered
                return False

        # Reset counter
        open_count = 0

        # Traverse from right to left, treating '*' as ')'
        for i in range(len(s) - 1, -1, -1):
            c = s[i]
            if c == ')' or c == '*':
                open_count += 1
            else:
                open_count -= 1

            if open_count < 0:  # If at any point, more '(' than ')' or '*' are encountered
                return False

        return True  # If all conditions are satisfied

File: 0678-valid-parenthesis-string/0678-valid-parenthesis-string.cpp
class Solution {
public:
    bool checkValidString(string s) {
        int open = 0, closed = 0;

        // Traverse from left to right, treating '*' as '('
        for (char c : s) {
            if (c == '(' || c == '*')
                open++;
            else
                open--;

            if (open < 0) // If at any point, more ')' than '(' or '*' are encountered
                return false;
        }

        // Reset counters
        open = 0;
        
        // Traverse from right to left, treating '*' as ')'
        for (int i = s.length() - 1; i >= 0; i--) {
            if (s[i] == ')' || s[i] == '*')
                open++;
            else
                open--;

            if (open < 0) // If at any point, more '(' than ')' or '*' are encountered
                return false;
        }

        return true; // If all conditions are satisfied
    }
};